File: C++/Trabalho_POO_parte_2/Simulador.cpp
#include "Simulador.h"

bool Simulador::verificar_id_existe(int id)
{
    for (Veiculo *veiculo : garagem)
        if (veiculo->get_id() == id)
            return true;
    return false;
}
Simulador::Simulador(int capacidade)
{
    capacidade_veiculos = capacidade;
    quantidade_veiculos = 0;
    std::cout << "SIMULAÇÃO COMEÇOU" << endl;
}

void Simulador::incluir_veiculo(char tipo)
{
    if (capacidade_veiculos == quantidade_veiculos)
        std::cout << "Capacidade de veículos está cheia" << endl;
    else
    {
        int id = rand() % 1000;
        while (verificar_id_existe(id))
            id = rand() % 1000;

        Veiculo *veiculo;
        if (tipo == 'B')
            veiculo = new Bicicleta(id, 2);
        else if (tipo == 'M')
            veiculo = new Motocicleta(id, 4);
        else if (tipo == 'C')
            veiculo = new Carro_Passeio(id, 4);
        else if (tipo == 'E')
            veiculo = new Carro_Esportivo(id, 4);

        garagem.push_back(veiculo);
        quantidade_veiculos = garagem.size();
        std::cout << veiculo->to_string() << endl;
        std::cout << "Quantidade de veículos na pista agora é " << quantidade_veiculos << endl;
    }
}

void Simulador::remover_veiculo(int id)
{
    for (int i = 0; i < garagem.size(); i++)
        if (garagem[i]->get_id() == id)
        {
            delete garagem[i];
            garagem.erase(garagem.begin() + i);
            quantidade_veiculos--;
            std::cout << "Veículo com ID " << id << " foi excluído da pista" << endl;
            std::cout << "Quantidade de veículos na pista agora é " << quantidade_veiculos << endl;
            break;
        }
    std::cout << "Não existe nenhum veículo com ID " << id << " na pista para ser excluído" << endl;
}

void Simulador::abastecer_veiculo(int id)
{
    for (Veiculo *veiculo : garagem)
        if (veiculo->get_id() == id)
        {
            Veiculo_Motorizado *motorizado = dynamic_cast<Veiculo_Motorizado *>(veiculo);
            if (motorizado != nullptr)
            {
                float quantidade_gasosa;
                std::cout << "Quantos litros de gasolina você quer colocar no tanque: ";
                cin >> quantidade_gasosa;
                motorizado->abastecer(quantidade_gasosa);
                break;
            }
            else
            {
                std::cout << "Bicicleta não pode ser abastecida" << endl;
                break;
            }
        }
    std::cout << "Não existe nenhum veículo com ID " << id << " na pista para ser abastecido" << endl;
}

void Simulador::mover_veiculo(int id)
{
    for (Veiculo *veiculo : garagem)
        if (veiculo->get_id() == id)
        {
            veiculo->mover();
            break;
        }
    std::cout << "Não existe nenhum veículo com ID " << id << " na pista para ser movido" << endl;
}

void Simulador::mover_veiculos_tipo(char tipo)
{
    for (Veiculo *veiculo : garagem)
        if (veiculo->get_tipo() == tipo)
            veiculo->mover();
}

void Simulador::mover_todos_veiculos()
{
    for (Veiculo *veiculo : garagem)
    {
        veiculo->mover();
    }
}

void Simulador::mostrar_dados_veiculos_tipo(char tipo)
{
    for (Veiculo *veiculo : garagem)
    {
        if (veiculo->get_tipo() == tipo)
        {
            std::cout << veiculo->to_string() << endl;
        }
    }
}

void Simulador::mostrar_dados_todos_veiculos()
{
    for (Veiculo *veiculo : garagem)
    {
        std::cout << veiculo->to_string() << endl;
    }
}

void Simulador::veiculo_mexer_pneu(int id)
{
    for (Veiculo *veiculo : garagem)
    {
        if (veiculo->get_id() == id)
        {
            int id_pneu = 0;
            for (string roda : veiculo->get_rodas())
            {
                std::cout << "Pneu " << id_pneu << " está " << roda << endl;
                id_pneu++;
            }

            int pneu_escolhido;
            std::cout << "Qual pneu você quer mexer? ";
            cin >> pneu_escolhido;

            if (pneu_escolhido >= 0 && pneu_escolhido < veiculo->get_quantidade_rodas())
            {
                int opcao;
                std::cout << "Esvaziar(1) ou encher(2)? ";
                cin >> opcao;

                if (opcao == 1)
                {
                    veiculo->esvaziar_pneu(pneu_escolhido);
                }
                else if (opcao == 2)
                {
                    veiculo->calibrar_pneu(pneu_escolhido);
                }
                else
                {
                    std::cout << "Digite 1 para esvaziar ou 2 para encher" << endl;
                }
            }
            else
            {
                std::cout << "Não existe esse pneu no veículo com ID " << id << endl;
            }

            return;
        }
    }
    std::cout << "Não existe nenhum veículo com ID " << id << " na pista para mexer no pneu" << endl;
}

void Simulador::calibrar_veiculos_tipo(char tipo)
{
    for (Veiculo *veiculo : garagem)
    {
        if (veiculo->get_tipo() == tipo)
        {
            for (int i = 0; i < veiculo->get_quantidade_rodas(); i++)
            {
                veiculo->calibrar_pneu(i);
            }
        }
    }
}

void Simulador::imprimir_pista()
{
    for (Veiculo *veiculo : garagem)
    {
        veiculo->desenhar();
    }
}

void Simulador::gravar_arquivo()
{
    if (garagem.size() > 0)
    {
        ofstream file("veiculos.bin", ios::binary);
        int tamanho = garagem.size();
        file.write(reinterpret_cast<char *>(&tamanho), sizeof(tamanho));
        for (Veiculo *veiculo : garagem)
        {
            veiculo->escrever_arquivo(file);
        }
        file.close();
        std::cout << "Veículos gravados com sucesso" << endl;
    }
    else
    {
        std::cout << "Não há veículos para serem gravados" << endl;
    }
}

void Simulador::ler_arquivo()
{
    ifstream file("veiculos.bin", ios::binary);
    if (file)
    {
        int tamanho;
        file.read(reinterpret_cast<char *>(&tamanho), sizeof(tamanho));
        for (int i = 0; i < tamanho; i++)
        {
            Veiculo *veiculo = Veiculo::ler_arquivo(file);
            garagem.push_back(veiculo);
        }
        file.close();
        std::cout << "Veículos lidos com sucesso" << endl;
    }
    else
    {
        std::cout << "Nenhum arquivo encontrado" << endl;
    }
}

File: Python/Trabalho_POO_parte_2/Simulador.py
# classes
from Veiculo import *
from Veiculo_Motorizado import *
from Bicicleta import *
from Motocicleta import *
from Carro_Passeio import *
from Carro_Esportivo import *

# biblioteca
from random import randint
import pickle
import os


class Simulador:
    # Construtor
    def __init__(self, capacidade: int) -> None:
        self.__capacidade_veiculos: int  = capacidade
        self.__quantidade_veiculos: int  = 0
        self.__garagem: list = []
        print("\n SIMULAÇÃO COMEÇOU \n")
  
    def incluir_veiculo(self, tipo: str) -> None:
            if self.__capacidade_veiculos == self.__quantidade_veiculos:
                print("Capacidade de veiculo está cheia")
            else:
                ids_existentes = set(veiculo.get_id() for veiculo in self.__garagem)
                id = None
                while id is None or id in ids_existentes:
                    id = int(randint(0, 1000))
                if tipo == 'B':
                    veiculo = Bicicleta(id, 2)
                elif tipo == 'M':
                    veiculo = Motocicleta(id, 2)
                elif tipo == 'C':
                    veiculo = Carro_Passeio(id, 4)
                elif tipo == 'E':
                    veiculo = Carro_Esportivo(id, 4)     
                self.__garagem.append(veiculo)
                self.__quantidade_veiculos = len(self.__garagem)
                print(veiculo)
                print(f"Quantidade de veiculos na pista agora é {self.__quantidade_veiculos}")

    def remover_veiculo(self, id: int) -> None:
        for i in range(len(self.__garagem)):
            if self.__garagem[i].get_id() == id:
                del self.__garagem[i]
                self.__quantidade_veiculos -= 1
                print(f"Veiculo com id {id} foi excluido da pista")
                print(f"Quantdade de __garagem na pista agora é {self.__quantidade_veiculos}")
                return
        print(f"Não existe nenhum veiculo com ID {id} na pista para que seja excluido")

    def abastecer_veiculo(self, id) -> None:
        for veiculo in self.__garagem:
            if veiculo.get_id() == id:
                if isinstance(veiculo, Veiculo_Motorizado):
                    quantidade_gasosa= float(input("Quantos litros de dado_float você quer colocar no tanque: "))
                    veiculo.abastecer(quantidade_gasosa)
                    return
                else:
                    print("Bicicleta não pode ser abastecida")
                    return
        print(f"Não existe nenhum veiculo com ID {id} na pista para que possa abastecer")

    def mover_veiculo(self, id) -> None:
        for veiculo in self.__garagem:
            if veiculo.get_id() == id:
                veiculo.mover()
                return
        print(f"Não existe nenhum veiculo com ID {id} na pista para que possa movelo")

    def mover_veiculos_tipo(self, tipo) -> None:
        if tipo in ["E", "B", "M", "C"]:
            for veiculo in self.__garagem:
                if veiculo.get_tipo() == tipo:
                    veiculo.mover()
                    return
        else:
            print(f"Não existe nenhum veiculo com tipo {tipo} na pista para que possa movelo")

    def mover_todos_veiculos(self) -> None:
        for veiculo in self.__garagem:
            veiculo.mover()

    def mostrar_dados_veiculos_tipo(self, tipo) -> None:
        if tipo in ["E", "B", "M", "C"]:
            for veiculo in self.__garagem:
                if veiculo.get_tipo() == tipo:
                    print(veiculo)
        else:    
            print(f"Não existe nenhum tipo {tipo} nesse Simulador")

    def mostrar_dados_todos_veiculos(self) -> None:
        for veiculo in self.__garagem:
            print(veiculo)

    def veiculo_mexer_pneu(self, id) -> None:
        for veiculo in self.__garagem:
            if veiculo.get_id() == id:
                id_pneu = 0 
                for roda in veiculo.get_rodas(): 
                    print(f"pneu {id_pneu} esta {roda}")   
                    id_pneu +=1
                while True:                
                    id_pneu = int(input("qual pneu vc quer mexer"))
                    if -1 < id_pneu and id_pneu < veiculo.get_quantidade_rodas():
                        while True:
                            try:
                                opcao = int(input("esvaziar(1) ou encher(2): "))
                            except ValueError:
                                print("Opção tipo de dados invalido'")
                            if opcao == 1:
                                veiculo.esvaziar_pneu(id_pneu)
                                break
                            elif opcao == 2:
                                veiculo.calibrar_pneu(id_pneu)
                                break
                            else:
                                print("Digite 1 para esvaziar ou 2 para encher")
                        break
                    else:
                        print(f"Não existe esse pneu no veiculo com ID {id}")
        print(f"Não existe nenhum veiculo com ID {id} na pista para que possa mexer no pneu dele")

    def calibrar_veiculos_tipo(self, tipo) -> None:
        if tipo in ["E", "B", "M", "C"]:
            for veiculo in self.__garagem:
                if veiculo.get_tipo() == tipo:        
                    for i in range(veiculo.get_quantidade_rodas()):
                        veiculo.calibrar_pneu(i)
                    break
        else:
            print(f"Não existe nenhum veículo com ID {id} na pista para que possa mexer nos pneus dele")
        
    def imprimir_pista(self) -> None:
        for veiculo in self.__garagem:
                veiculo.desenhar()
        
    def gravar_arquivo(self) -> None:
        if len(self.__garagem) != 0:
            with open('veiculos.txt', mode='wb') as file:
                pickle.dump(len(self.__garagem), file)
                for veiculo in self.__garagem:
                    pickle.dump(veiculo, file)
            print("Veículos gravados com sucesso")
        else:
            print("Não há veículos para serem gravados")

    def ler_arquivo(self) -> None:
        if os.path.isfile('veiculos.txt'):
            for i in range(len(self.__garagem)):
                del self.__garagem[i]
            
            with open('veiculos.txt', 'rb') as file:
                for _ in range(pickle.load(file)):
                    self.__garagem.append(pickle.load(file))
            print("Veículos lidos com sucesso!")
        else:
            print("Nenhum arquivo encontrado")
    